---
layout: post
title: Java核心技术第7章异常、断言和日志
category: Java
tags: [《Java核心技术》]
description:
---
## Java异常

* 在Java中，异常对象都是派生于Throwable类的一个实例。
* 所有的异常都是由Throwable继承而来，但在下一层立即分解为两个分支：Error和Exception。
* Error列层次结构描述了Java运行时系统的内部错误和资源耗尽错误。
* Exception又分解为两个分支：RuntimeException和其他异常。由程序错误导致的异常属于RuntimeException；程序没问题，但由于I/O错误导致的异常属于其他异常。
* RuntimeException包括：错误的类型转化，数组访问越界，访问null指针
* 不是派生于RuntimeException的异常包括：试图在文件尾部后面读取数据，试图打开一个不存在的文件。
* 如果出现RuntimeException异常，则一定是你的问题。
* 派生于Error类和RuntimeException类的所有异常为非受查异常，其他的异常称为受查异常。
* 一个方法必须声明所有可能抛出的受查异常，而非受查异常要么不可控（Error）要么应该避免，发生（RuntimeException）。如果一个方法没有声明所有可能发生的受查异常，编译器就会发出一个错误消息。
* 如果在子类中覆盖了超类的一个方法， 子类方法中声明的受查异常不能比超类方法中声明的异常更通用 （也就是说， 子类方法中可以抛出更特定的异常， 或者根本不抛出任何异常）。特别需要说明的是， 如果超类方法没有抛出任何受查异常， 子类也不能抛出任何受查异常。
* 抛出异常：找到一个合适的异常类，创建这个类的异常对象，将对象抛出。

      String readData(Scanner in) throws EOFException{
        if(...){
          throw new EOFException();
        }
      }

* 捕获异常：使用try/catch语句块。如果在 try语句块中的任何代码抛出了一个在 catch 子句中说明的异常类， 那么
程序将跳过 try语句块的其余代码。并执行 catch 子句中的处理器代码。

      try
      {
        ....
      }
      catch(ExceptionType e){
        ....
      }

* 通常， 应该捕获那些知道如何处理的异常， 而将那些不知道怎样处理的异常继续进行传递。
* 要想获得对象的更多信息， 可以试着使用 e.getMessage（）
* finally：由于try中抛出异常后会跳过其余代码，可以会对资源释放产生影响，使用final语句保证无论是否抛出异常，finally中的代码始终被执行。执行顺序：try抛出异常前的代码，catch中的代码，finally中的代码，如果抛出的异常不在catch的捕获范围内，则try抛出异常前的代码执行完后直接执行finally。

      try
      {
        ....
      }
      catch(IOException e)
      {
        ....
      }
      finally
      {
        closefile;
      }

## 断言
断言机制允许在测试期间向代码中插入一些检査语句。当代码发布时，这些插人的检测语句将会被自动地移走。    
Java语言引入关键字assert，这个关键字有两种形式

      assert 条件；
      和
      assert 条件：表达式；
